src/main/java/org/wahlzeit_revisited/repository/AbstractRepository.java

```diff
@@ -0,0 +1,175 @@
+package org.wahlzeit_revisited.repository;
+
+import org.wahlzeit.services.SysLog;
+import org.wahlzeit_revisited.database.DatabaseConnection;
+import org.wahlzeit_revisited.database.SessionManager;
+import org.wahlzeit_revisited.model.Photo;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public abstract class AbstractRepository<T extends Persistent> implements Repository<T> {
+
+    /*
+     * template methods
+     */
+
+    protected abstract PersistentFactory<T> getFactory();
+
+    protected abstract String getTableName();
+
+    /*
+     * Repository contract
+     */
+
+    public Optional<T> findById(Long id) throws SQLException {
+        assertIsNonNullArgument(id);
+
+        String query = String.format("SELECT * FROM %s WHERE id = ?", getTableName());
+        PreparedStatement stmt = getReadingStatement(query);
+        stmt.setLong(1, id);
+
+        T persistent = null;
+        try (ResultSet resultSet = stmt.executeQuery()) {
+            if (resultSet.next()) {
+                persistent = getFactory().createPersistent(resultSet);
+            }
+        }
+
+        return Optional.ofNullable(persistent);
+
+    }
+
+    public List<T> findAll() throws SQLException {
+        String query = String.format("SELECT * FROM %s", getTableName());
+        PreparedStatement stmt = getReadingStatement(query);
+
+        List<T> result = new ArrayList<>();
+        try (ResultSet resultSet = stmt.executeQuery()) {
+            while (resultSet.next()) {
+                T persistent = getFactory().createPersistent(resultSet);
+                result.add(persistent);
+            }
+        }
+
+        return result;
+    }
+
+    public T insert(T toInsert) throws SQLException {
+        assertIsNonNullArgument(toInsert);
+        assertNonPersistedObject(toInsert);
+
+        String query = String.format("INSERT INTO %s DEFAULT VALUES RETURNING id", getTableName());
+        PreparedStatement stmt = getReadingStatement(query);
+
+        // Extract id from RETURNING statement
+        long persistedId;
+        try (ResultSet returningSet = stmt.executeQuery()) {
+            returningSet.next();
+            persistedId = returningSet.getLong(1);
+            toInsert.setId(persistedId);
+        }
+
+        // Get empty entity row
+        query = String.format("SELECT * FROM %s WHERE id = ?", getTableName());
+        stmt = getUpdatingStatement(query);
+        stmt.setLong(1, persistedId);
+
+        // Write internal structure in resultSet and start transaction
+        try (ResultSet returningSet = stmt.executeQuery()) {
```

Comment:
Is there a need for an additional try/catch block? 

If this operation fails due any reasons, there are orphans rows inside the db, which could mess up the application semantics. 

Eg a new `User` get's inserted:
When the db-connection get's closed between the `INSERT`-Call and and the `returningSet.updateRow()`; there is no way to delete that empty row, but there is the chance, that with the `Repository::findAll()`, this `User` get's instantiated. But this operation will fail as `email_address`-row is `null` and the `EmailAddress` class, will assert valid/nonNull email-strings.